reactor-core/src/test/java/reactor/core/publisher/FluxBufferTest.java

```diff
@@ -661,4 +668,96 @@ public void discardOnErrorOverlap() {
 		            .verifyThenAssertThat()
 		            .hasDiscardedExactly(1, 2, 3, 3); //we already opened a 2nd buffer
 	}
+
+    @ParameterizedTestWithName
+    @CsvSource({
+        "1|2,     1|2,        ",
+        "1|1|1,   1,       1|1",
+        "1|1|2,   1|2,       1",
+        "1|2|1,   1|2;1,      ",
+        "1|2|1|3, 1|2;1|3,    ",
+        "1|1|2|3, 1|2;3,     1",
+        "2|1|1|3, 2|1;1|3,    "
+    })
+    public void bufferExactSupplierUsesSet(String input, String output, @Nullable String discard) {
+        List<Set<Object>> outputs = Arrays.stream(output.split(";"))
+            .map(it -> Arrays.<Object>stream(it.split("\\|")).collect(Collectors.toSet()))
+            .collect(Collectors.toList());
+
+        StepVerifier.Assertions assertions = Flux.just(input.split("\\|"))
+            .<Collection<Object>>buffer(2, HashSet::new)
+            .as(it -> StepVerifier.create(it, outputs.size()))
+            .expectNextSequence(outputs)
+            .expectComplete()
+            .verifyThenAssertThat(Duration.ofSeconds(2));
+
+        if (discard == null) {
+            assertions.hasNotDiscardedElements();
+        } else {
+            assertions.hasDiscardedExactly((Object[]) discard.split("\\|"));
+        }
+    }
+
+    @ParameterizedTestWithName
+    @CsvSource({
+        "1|2,       1|2,          ",
+        "1|1|1,     1,         1|1",
+        "1|1|2,     1|2,         1",
+        "1|2|1,     1|2,          ",
+        "1|2|1|3,   1|2;3,       1",
+        "1|2|1|1|3, 1|2;1|3,     1",
+        "1|1|2|3,   1|2,         1",
```

Comment:
What happened to 3? It should be discarded or delivered, but not missing.

Reply:
More on that in the overall comment.

Reply:
Ah good catch; The test only requests as many buffers as explicitly stated here, so the `3` was just unaccounted for